### DISCUSSION TITLE
contextual vs non-contextual quantities

### DISCUSSION THREAD

**Discussion by Toby-Broom** [qudt/qudt-public-repo#1079] [General]
I think @fkleedorfer brought the point about the usability/experiance from the user perspective as a fair one that having some more abstract units makes less easy to use.  I think these are domain specific so we can model these lower down, which is what we have being doing in our projects.

In a more general sense I think that you would need to describe context in your data anyway so while the unit count maybe a little strange alone its never out of context.

like the meter example, I can measure a cells (has diameter), persons (has height), frames (has width) in this quantity also so many things could be expanded to be contextual which would be a lot of new things in qudt?

**Comment by Toby-Broom** [comment]
For the people that wanted to use CFU for example they could create their own instances and link back to QUDT?

**Reply by fkleedorfer** [reply]
Definitely. Although it requires a little bit of care: the applicableUnits triples are autogenerated for all units in QUDT. These must be added along with the unit for optimal user satisfaction.

**Comment by fkleedorfer** [comment]
We are currently thinking about domain specific or industry specific profiles.(https://github.com/qudt/qudt-public-repo/discussions/1077) Maybe the problem of domain/context specific units could be addressed with that mechanism. 

The domain specific profiles could contain mappings from units used in the domain to the unit/quantitykind combination that represents it in QUDT. E.g. in the biology profile, there may ve a mapping CFU --> qudt:NUM/qk:ColonyFormingNumber (or whatever the qk is for cfu).

Moreover, context specific units (such as cfu) could be added to the profile, in which case the above mapping might be unnecessary as the core QUDT unit/qk can be referenced from that context specific unit (and maybe this would be the best way to define such mapping)

The profiles could invite contributions from users in the domain for those mappings/unit definitions, and maybe even for widely used quantities, if they can be defined for the domain.

**Reply by Toby-Broom** [reply]
I think this is in line with what I was thinking above

**Comment by Toby-Broom** [comment]
I think it should be handled with 2 triples since it a robust handling of all situations.  You describe what you're counting and how much there is of the thing that you count. e.g.

The definition applies to entities B which should always be indicated by a subscript or in.  parentheses, e.g. nB or n(B). When the chemical composition is written out, parentheses should be used, n(O2).

[https://doi.org/10.1039/9781847557889](url) Chapter 2.10

This would be for the amount of oxygen, which is a count of the molecule.

We often display something slightly different to how the data is stored to improve the user experiance.

**Reply by fkleedorfer** [reply]
This would mean using such a string (e.g. `"kg(CO2)"`) for the `qudt:symbol` of the context-dependent unit, which would be something like `qudt:KiloGM_O2` (if the naming convention allowes it). The problem we have been trying to avoid there is that this leads to a combinatorial explosion of units (all relevant scalings of all relevant units for each chemical compound). By using a different QuantityKind for different chemical compounds, the vocabulary size grows only linearly with number of chemical compounds. This problem would not go away if we only defined such units in a profile. 

If it's just about the symbol, would it be a problem to add a, say, `qudt:symbolSuffix` to the QuantityKind that is specific to the chemical compound? Thus, we could annotate the CO2 related quantitykind with `qudt:symbolsSuffix "(CO2)"`, and applications that want to generate quantitykind-specific labels can use that information?

**Reply by Toby-Broom** [reply]
I meant from a ontology perspective, I would split things up:

CO2 has a quantity value
quantiity value is x kg

or for the bio things

Colonies have a count/number
count/number is 100

In the UI I would use the symbol rules and put 10 kg(CO2)

**Reply by fkleedorfer** [reply]
So the suggested additional triple would be a viable solution for this use case, right?

**Reply by Toby-Broom** [reply]
I think this is general solution for all, there is always a context to a quantity.  In the end also not many people are interacting directly with the ontology.  I my previous company we stored all the data in base SI units but people still wanted rpm or °C on the UI because its a better user experiance.

**Comment by steveraysteveray** [comment]
After following this discussion and thinking about some existing context-dependent units in our current units vocabulary, I have the following proposal:

1. Move existing units that are context-dependent to a new graph named something like http://qudt.org/$$QUDT_VERSION$$/vocab/contextualunit. Candidate entries include compound units that are not really units but are in common use, such as:
- unit:BEAT-PER-MIN as a measure of pulse, or heart rate
- unit:BREATH-PER-MIN as a measure of respiration rate rate
- unit:CASES-PER-KiloINDIV-YR for morbidity rate (probably should use singular unit:CASE)
- unit:DEATHS-PER-KiloINDIV-YR for mortality rate (again, probably unit:DEATH)
- unit:CYC-PER-SEC, which is qudt:exactMatch unit:HZ
- unit:FRAME-PER-SEC
- unit:SAMPLE-PER-SEC
- unit:EUR-PER-KiloW-HR, where EUR is a CurrencyUnit with prefix cur:
- unit:CFU, for Colony Forming Unit, a unit for Microbial Formation that is equivalent to unit:NUM
- unit:GM_Carbon-PER-M2-DAY, where the unit is being conflated with the measurement
...as well as other domain-specific units such as:

This last example is nicely addressed in the recent [publication in Nature](https://www.nature.com/articles/s41597-025-04587-8) in which the authors point  out:

"_A particular difficulty is the conflation of units with other aspects of measurements, for example, an observation that combines nouns with a unit. For example, “milligrams of nitrogen” may be the name of a measurement or in the QUDT model, a Quantity but the unit is simply milligrams. There are advantages to separating units and measurements (quantities)._"

2. Moving these units to a different graph means we can avoid deprecating any of them, and thus reduce disruption to current users.

3. None of these compound units would require factor units for all components. In other words, we would not include unit:BEAT or unit:BREATH. 

4. Each of these new entries would have two triples that tie it to a Unit and a QuantityKind, e.g.
```
unit:BREATH-PER-MIN skos:related unit:NUM-PER-MIN  (or skos:broader could be used)
unit:BREATH-PER-MIN qudt:hasQuantityKind quantitykind:RespiratoryRate
```

This approach cleans up the QUDT units vocabulary graph, but still allows provision of terms that users sometimes think of as units and points them to the underlying "context-free" QUDT unit. QUDT.org could curate common ones, and the user community could provide additional graphs containing terms of art as extensions.

Thoughts?

**Comment by jhodgesatmb** [comment]
The only issue I see with this suggestion is where to draw the line of what is considered contextual. Taking the devil's advocate position, the same argument could be made for any 'qualified' units (i.e., with underscores after the unit). It is really a matter or perspective or threshold for what defines 'contextual'. If we don't define the threshold in such a way that it is clear and unambiguous we would really be pushing the problem down the road.

What is the rationale for this proposal? The discussion topic has roamed a bit, from contextual units, to profiles, and back again. The point that @toby and @florian made is that context opens up a virtually infinite number of units. So far, we have allowed domain experts to submit contextual units, and some standards bodies also have contextual units (at least, the qualifier types). Is that where profiles enter the discussion?

**Comment by steveraysteveray** [comment]
I agree that we will need to try to be as clear as possible. Some "underscored" units, I would argue, are not contextual in the sense used above, such as GAL_US vs GAL_UK, or DEG_C vs DEG_F. They are just different variants of an ambiguously defined unit. But your point is well taken.

The rationale for this became clear to me when looking at the factorization work that is underway. The first 7 examples I listed above exposed the fact that the component units were undefined (such as BEAT, BREATH, CASES, etc.) and are really not units, just countable phenomena. Rather than requiring the components to be defined, I feel it is better to simply define the compound units but not require full factorization.

The bigger picture I see is one that uses both contextual unit graphs (contributed by users plus one curated by QUDT.org) AND profiles that could be defined as the union of one or more contextual unit graphs and a filtered subset of the main unit graph.

**Reply by jhodgesatmb** [reply]
I agree that many of the 'qualified' units aren't contextual. What I was saying is that some of them are, and that extends the number of units affected by your proposal. Examples are: CAL_15_DEG_C, unit:CentiM_H2O, unit:IN_H2O_39dot2DEG_F, etc. One could argue that any material type or temperature specialization is a context. That argument is easily extended in any direction.

**Comment by steveraysteveray** [comment]
Yes, I think those would all move over. I agree that any material type or temperature specialization like these are contextual.

**Reply by fkleedorfer** [reply]
I'd like to emphasize that more files mean more work. Especially in the build process, If we need to have certain steps done to multiple files in a similar way, it adds to the complexity. I was kind of hoping we could put the currencies back into the units file in src - after all, we can identify them by their class.... and maybe we could just invent a new unit class 'ContextualUnit' that we treat a little differently during the build?

**Reply by fkleedorfer** [reply]
... It's not necessarily better for users to have to figure out which N of the M units files they need for their application.

**Reply by steveraysteveray** [reply]
Are you saying that all the units could live in the single file, with CurrencyUnits declared as such, and ContextualUnits similarly declared? Then the inference rules would skip over those when calculating factor units, right? Or maybe they would just infer factor units when the component unit exists since it's not usually... I could work on that query.

I would need to tweak the publication scripts if we merged all units - not impossible. We could discuss whether currencies should retain cur: prefix.

**Reply by fkleedorfer** [reply]
Yes, along those lines, and if really needed, we can split them up in multiple files in the target

**Reply by steveraysteveray** [reply]
That could work, but even better, let's figure out how to build profiles. The default would be "everything" - all units in one graph, all quantitykind in one graph, etc.

What I think we need is a way to invoke a build, presumably locally, that will do a combination of the following things:
1. Filter out things by means of some predefined SPARQL query. This could include FILTER statements like excluding currency units.
2. Add things stored in extension graphs that don't really belong in the main graphs. The extension graphs could include domain-specific units, plus contextual units.

I'm not sure what is the most elegant way to select arbitrary combinations of the above. 
-P parameters in Maven?

**Reply by steveraysteveray** [reply]
And to your point, @fkleedorfer, certain users have made clear to me that they at least know parts of QUDT that they definitely do NOT want cluttering up their workspace.

**Reply by steveraysteveray** [reply]
Thinking of the user's perspective, it seems like the Maven -P mechanism might do the trick. The "P" stands for profile, after all. We could have options like:
- mvn -Pall install, which would work much like the current 'mvn install'
- mvn -Pcore install, where filtered versions of the vocabularies are staged for the install, and would not have currencies or the context-dependent units or quantity kinds.
- mvn -Pbiology install, which would filter out concepts clearly out of scope, such as astronomical, electromagnetic, etc., plus adding biology-context-dependent concepts.
- mvn -P<any community-defined profile> install, ...
- etc.

Still to be determined how best to design the back end of such functionality.

Thoughts?

**Reply by fkleedorfer** [reply]
If we decide that self-generated profiles are sufficient for the 'profiles' feature, that would make things easy. As long asit is just a question of whether x is in profile a or not, no big deal, but I think there are going to be multiple dimensions along which people are going to want to choose. Then, a distribution zip becomes a tangled mess..

**Comment by Toby-Broom** [comment]
Sorry for late reply was sick, what sort of profiles would you have by domains like economics, biology.?  I do like the idea of having smaller items, it seems tough when we have an ontology or vocab with 1000's of items when ofen we only need less than 100 items to decribe a problem, so it nice to have modularity, if we ever needed everything we can load all of the modules, the solutions seem more to a domain, so it fits to take the module for a domain. 

My Endotoxin units were prescibed by US Pharmacopeia so recognised in some manner by domain experts.

We are using the IOF for measurements (e.g. https://spec.industrialontologies.org/ontology/202401/core/Core/MeasurementProcess) so we already have the higher level item that the nature article talks about.   You can descibe things like Jack showed properly since the measurement process outputs more than one quantlity value the temperature and the energy.  There is also materials included so we can attach the measurement to a material so we descibe what object that is being measured. I do see what the paper descibes in my work that how a person would descibe something e.g. "phosphate concentration in soil pore water" does seem to lend itself to something more ontologically complex, I feel that it has value from an answering questions perspective however if you had the question what do phosphates do then it would be hard to tell if you only had the concept of a "phosphate concentration".

I'm happy to contribute what I can.

**Comment by steveraysteveray** [comment]
Revisiting this issue since some relevant things have now been automated. Specifically, we now support ContextualUnit in the units graph by declaring such unit to be of type qudt:Unit and qudt:ContextualUnit. Further, as of Release 3.1.7, the ContextualUnit type propagates between contextual compound units and supporting contextual component units. The applicableUnit relation still works:
- A specialized contextual unit can point to a specialized quantity kind (which would have a skos:broader relation to a more general quantity kind), or can just point to a general quantity kind. Either approach seems fine to me.
- If we wanted to restore unit:CFU (colony forming unit), it could either point to quantitykind:ColonyFormingNumber or something like that, or could just point to quantitykind:Count. If the combination of unit and quantitykind is important to an application, and each needs to be unique to that discipline, then use a specialized quantity kind like quantitykind:ColonyFormingNumber.

Am I missing anything? Users can filter the QUDT vocabularies as desired, or can use profiles if we iron out that functionality.

**Comment by Toby-Broom** [comment]
Do you have a example?  I have another project that uses CFU :)

**Reply by steveraysteveray** [reply]
I don't have examples - you are the domain expert here! 

Now that we are armed with the ContextualUnit pattern, I'm willing to endorse restoring unit:CFU and even unit:ENDOTOXIN. (I disagree with your assertion in #1041 that you cannot generalize to MOL. Even with a mixture of things, you can still talk about a mole of that mixture.) So sure, unit:ENDOTOXIN-PER-L skos:broader unit:NUM-PER-L, or maybe unit:MOL_ENDOTOXIN-PER-L skos:broader unit:MOL-PER-L, and unit:MOL_ENDOTOXIN-PER-L qudt:hasQuantityKind quantitykind:Concentration or quantitykind:AmountOfSubstanceConcentration (they mean the same thing).

**Reply by Toby-Broom** [reply]
I mean an example of qudt:ContextualUnit does it follow same pattern as  qudt:Unit?

I agree you can have a mol of anything, I think my point was that a mol is just a number since its the count of particles, e.g. N0 = 6.02214076×1023

**Reply by steveraysteveray** [reply]
Ah, I understand now. Please see this wiki page on [Contextual Units](https://github.com/qudt/qudt-public-repo/wiki/ContextualUnits) that is referred to in our [Unit Vocabulary Submission Guidelines](https://github.com/qudt/qudt-public-repo/wiki/Unit-Vocabulary-Submission-Guidelines). You can see our listing of currently defined ContextualUnits [here](https://yasgui.org/#query=PREFIX+dcterms%3A+%3Chttp%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%3E%0APREFIX+sou%3A+%3Chttp%3A%2F%2Fqudt.org%2Fvocab%2Fsou%2F%3E%0APREFIX+unit%3A+%3Chttp%3A%2F%2Fqudt.org%2Fvocab%2Funit%2F%3E%0APREFIX+skos%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2004%2F02%2Fskos%2Fcore%23%3E%0APREFIX+qudt%3A+%3Chttp%3A%2F%2Fqudt.org%2Fschema%2Fqudt%2F%3E%0APREFIX+vaem%3A+%3Chttp%3A%2F%2Fwww.linkedmodel.org%2Fschema%2Fvaem%23%3E%0APREFIX+rdf%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0APREFIX+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0ASELECT+DISTINCT+*%0AWHERE+%7B%0A++%3Fu+a+qudt%3AContextualUnit+.%0A++FILTER+NOT+EXISTS+%7B%3Fu+qudt%3Adeprecated+true%7D+.%0A%7D&contentTypeConstruct=text%2Fturtle&contentTypeSelect=application%2Fsparql-results%2Bjson&endpoint=https%3A%2F%2Fwww.qudt.org%2Ffuseki%2Fqudt%2Fsparql&requestMethod=POST&tabTitle=Query+4&headers=%7B%7D&outputFormat=table).
